### PR TITLE
feat: implements heartbeat

### PR DESCRIPTION
**Pull request by f-erdi1990**
Added variables to guide cadence and signer ("deadMan") of switch as well as core heartbeat method.
Aliveness checks based on "read" method which returns aliveness. Technically, isAlive method can be deleted.

### CODE REVIEW

File: contracts/DeadMansSwitch.sol [hunk 1]
```diff
@@ -7,10 +7,44 @@ error DeadMansSwitchError();
 
 contract DeadMansSwitch {
     bool public isAlive;
+    address public deadMan;         // signer that can sign heartbeats
```
Comment (arein) [inline contracts/DeadMansSwitch.sol:10]: nit: consider using [onlyOwner](https://docs.openzeppelin.com/contracts/2.x/api/ownership) for this

File: contracts/DeadMansSwitch.sol [hunk 2]
```diff
@@ -7,10 +7,44 @@ error DeadMansSwitchError();
...
         console.log("Deploying a DeadMansSwitch with isAlive:", true);
         isAlive = true;
+        deadMan = _deadMan;
+        downCadence =  _downCadence;
+        cadence = _cadence;
+        startWindowBlock = block.number + _downCadence;
+        endWindowBlock = block.number + _cadence;
+    }
+
+    function heartbeat() external returns (bool) {
```
Comment (arein) [inline contracts/DeadMansSwitch.sol:34]: `onlyOwner` would be useful here

File: contracts/DeadMansSwitch.sol [hunk 3]
```diff
@@ -7,10 +7,44 @@ error DeadMansSwitchError();
...
+        require(msg.sender == deadMan, 'Dead Man Switch: Wrong Dead Man');
+        uint blockNo = block.number;
+        if (blockNo > endWindowBlock) {
+            revert ManIsUnfortunatelyAlreadyDead();
+        } else if (blockNo < startWindowBlock) {
+            revert SwitchNotReadyYet();
+        } else {
+        startWindowBlock = blockNo + downCadence;
```
Comment (arein) [inline contracts/DeadMansSwitch.sol:42]: I don't fully understand why a `startWindowBlock` is needed? why not always allow?

File: contracts/DeadMansSwitch.sol [hunk 4]
```diff
@@ -22,7 +56,21 @@ contract DeadMansSwitch {
         isAlive = _isAlive;
     }
 
+    function checkAliveness() public view returns (bool) {
```
Comment (arein) [inline contracts/DeadMansSwitch.sol:59]: optional: `pinch` is more in the lingo of the analogy

```suggestion
    function pinch() public view returns (bool) {
```

**Review by arein — APPROVED** [review]
It would be nice if you could cover this in the `./tests` which you can then execute via `yarn test`

Also nit: the existing tests now probably fail

File: contracts/DeadMansSwitch.sol (hunk 1 above)
Comment (f-erdi1990) [reply contracts/DeadMansSwitch.sol:10]: yup, can do

File: contracts/DeadMansSwitch.sol (hunk 2 above)
Comment (f-erdi1990) [reply contracts/DeadMansSwitch.sol:34]: why do I need only owner? I require msg.sender to be == deadMan (or onlyOwner in new terminology)... Not sure where onlyOwner would need to go

File: contracts/DeadMansSwitch.sol (hunk 3 above)
Comment (f-erdi1990) [reply contracts/DeadMansSwitch.sol:42]: Not Needed, optional argument to decrease window size. Just gives people more variability in how they set up their dead man's switch

**Pull request closed without merging**